src/SAML2/XML/saml/AbstractBaseIDType.php

```diff
@@ -0,0 +1,70 @@
+<?php
+
+declare(strict_types=1);
+
+namespace SAML2\XML\saml;
+
+use DOMElement;
+use SAML2\Constants;
+use SAML2\DOMDocumentFactory;
+use SAML2\XML\IDNameQualifiersTrait;
+use Webmozart\Assert\Assert;
+
+/**
+ * SAML BaseIDType abstract data type.
+ *
+ * @author Jaime Pérez Crespo, UNINETT AS <jaime.perez@uninett.no>
+ * @package simplesamlphp/saml2
+ */
+abstract class AbstractBaseIDType extends AbstractSamlElement
```

Comment:
I'm not sure it;s a good idea to mix several kinds of tuples in our class registry..

Reply:
Regarding (3).. Do we really need the extra param? Can't we use self::getClassName(self::class) ?

Reply:
(5) right now is creating a circular dependency, unless I've missed something in the plan..

Reply:
(6) I'm not sure it;s a good idea to mix several kinds of tuples in our class registry..

Reply:
Let's see:

> Regarding (2) and (3).. They seem to contradict each other..
Do we really need the extra param and/or the constant? Can't we use self::getClassName(self::class) ?

It's the opposite, they complement each other. If the class is not abstract, then you need to make sure an `xsi:type` is always available. The only way you could skip passing the `xsi:type` as an argument to the constructor would be to make it abstract and then use the extending class' name as the value for `xsi:type`.

> (5) right now is creating a circular dependency, unless I've missed something in the plan..

No, because `BaseID::fromXML()` always returns an object of class `BaseID` (which is the reason for it not being abstract). So `fromXML()` in an extending class would create an object of that class, based on the information available in the `BaseID` object. E.g.:

```php
public static function fromXML(DOMElement $e): MyID
{
    $baseid = BaseID::fromXML($e);
    $attr = $e->getAttribute('myattr');
    return new self(
        $baseid->getValue(),
        $baseid->getNameQualifier(),
        $baseid->getSPNameQualifier(),
        $attr
    );
}
```

> (6) I'm not sure it;s a good idea to mix several kinds of tuples in our class registry..

Yeah, I reckon I don't like that either. What about this?:

1. The `registerClass()` method would check if the given class extends `BaseID`, then register it with an index like `BaseID:whatever_value_we_set_on_xsi:type`.
2. We add another method explicitly searching the registry for `BaseID` extensions: `getCustomIDClass($xsitype)`.

I don't see any issue about both things coexisting in the same registry, since it's just an index that we control, but having another explicit method is probably a good idea.

Reply:
Would that make sense?

Reply:
I'm sorry, you've lost me on all bullets...

Reply:
Let me try to suggest an implementation, and let's see from there if it makes sense.

Reply:
Thanks, and sorry for me being a scatterbrain today

Reply:
No worries, I'm there myself most days anyway 😉

Reply:
We should start a club